internals/overlord/logstate/gatherer.go

```diff
@@ -169,40 +175,47 @@ func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.R
 // pullers on entryCh, and writes them to the client. It also flushes the
 // client periodically, and exits when the gatherer's tomb is killed.
 func (g *logGatherer) loop() error {
-	timer := newTimer()
-	defer timer.Stop()
+	flushTimer := newTimer()
+	defer flushTimer.Stop()
+
+	flushClient := func(ctx context.Context) {
+		// Mark timer as unset
+		flushTimer.Stop()
+		err := g.client.Flush(ctx)
+		if err != nil {
+			logger.Noticef("Cannot flush logs to target %q: %v", g.targetName, err)
+		}
+	}
 
 mainLoop:
 	for {
 		select {
 		case <-g.tomb.Dying():
 			break mainLoop
 
-		case <-timer.Expired():
-			// Mark timer as unset
-			timer.Stop()
-			err := g.client.Flush(g.clientCtx)
-			if err != nil {
-				logger.Noticef("Cannot flush logs to target %q: %v", g.targetName, err)
-			}
+		case <-flushTimer.Expired():
+			flushClient(g.clientCtx)
 
 		case entry := <-g.entryCh:
-			err := g.client.Write(g.clientCtx, entry)
+			err := g.client.AddLog(entry)
 			if err != nil {
 				logger.Noticef("Cannot write logs to target %q: %v", g.targetName, err)
+				continue
+			}
+			// Check if buffer is full
+			if g.client.NumBuffered() >= g.maxBufferedEntries {
+				flushClient(g.clientCtx)
 			}
-			timer.EnsureSet(g.bufferTimeout)
+			// Otherwise, set the timeout
+			flushTimer.EnsureSet(g.bufferTimeout)
```

Comment:
The comment says "Otherwise", but this is not in an `else`. Should it be?

Reply:
whoops, meant to put in a `continue`